compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3144,6 +3144,125 @@ static inline void loadArrayCmpSources(TR::Node *node, TR::InstOpCode::Mnemonic
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, loopStartLabel);
+   startLabel->setStartInternalControlFlow();
+
+   // main-loop
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec0Reg, indexReg, src1AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvb16x, node, vec1Reg, indexReg, src2AddrReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb_r, node, vec0Reg, vec0Reg, vec1Reg);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, resultLabel, condReg);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, indexReg, indexReg, 16);
+
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bdnz, node, loopStartLabel, condReg);
+   // main-loop end
+
+   // residue start
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, residueStartLabel);
+
+   generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, temp2Reg, returnReg, 56, 0xFF00000000000000);   
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src1AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec0Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src2AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec1Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb, node, vec0Reg, vec0Reg, vec1Reg);
+   // residue end
+
+   // result
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, resultLabel);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vec0Reg, vec0Reg, vec0Reg);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::vclzlsbb, node, temp2Reg, vec0Reg);
+   
+   // vec-eq-mask = -1 (if equal) | 0 (it not-equal)
+   generateShiftLeftImmediateLong(cg, node, tempReg, temp2Reg, 59);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sradi, node, tempReg, tempReg, 63);
+
+   // offset = matched-byte-count & ~(vec-eq-mask) + remainder & (vec-eq-mask)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, temp2Reg, temp2Reg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::AND, node, returnReg, returnReg, tempReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, temp2Reg);
+
+   // index = index + offset, if we need to return unmatched index, then we are done here 
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, indexReg, returnReg);
+   
+   if (!node->isArrayCmpLen())
+   {
+      // if it was a match then we need index -= 1 to get last index for compare
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, tempReg);
+
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, src1AddrReg, src1AddrReg, returnReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, src2AddrReg, src2AddrReg, returnReg);
```

Comment:
After discussing with Devin, I included zero length arrays in my sanity test case.

Reply:
I double checked, for zero length arraycmp, the generated code is never called (breakpoint didn't hit). (I used count=0 to compile and check the method). What is your suggestion, should I also cover the zero length anyways?

Reply:
Hmm, this looks like a strange problem. Looking into this more, you're right that we don't currently use this for 0 length compares. Furthermore, it seems that the old version also has problems with zero length compares. I've opened #6015 to start some discussion about what the semantics of this really should be. Since this isn't really regressing anything over the old version, this PR can probably go ahead without worrying about this and we can come back to this if the decision is that the semantics should be tightened up for lengths of 0.